### ISSUE TITLE
System.Net.Http not the same types for .Net461 when using System.Threading.Tasks.Extensions

### ISSUE THREAD

**Issue by rhyous** [dotnet/corefx#27907]
Visual Studio: 2017
Project: .Net4.6.1
Package:   <package id="System.Threading.Tasks.Extensions" version="4.4.0" targetFramework="net461" />

### Problem
.NET 4.6.1 doesn't see the System.Net.Http as the same types. I get type exceptions or missing method exceptions when trying to call code from System.Net.Http namespace. 

It seems to fail in many uses of the System.Net.Http namespace.
Once particular example, I have an IHttpClient interface and an HttpClientWrapper class to wrap HttpClient and allow for unit test. My unit tests don't work.

### Workaround

The csproj is pointing at the Netstandard2.0 version of the dll, which must have some bug that makes the System.Net stuff look like they are different types.

```
    <Reference Include="System.Threading.Tasks.Extensions, Version=4.1.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\..\packages\System.Threading.Tasks.Extensions.4.4.0\lib\netstandard2.0\System.Threading.Tasks.Extensions.dll</HintPath>
    </Reference>
```

Manually edit the project. Change it to point to the portable-net45+win8+wp8+wpa81.

```
    <Reference Include="System.Threading.Tasks.Extensions, Version=4.1.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\..\packages\System.Threading.Tasks.Extensions.4.4.0\lib\portable-net45+win8+wp8+wpa81\System.Threading.Tasks.Extensions.dll</HintPath>
    </Reference>
```

### Expectation
I shouldn't have to manually edit my project every time I update System.Threading.Tasks.Extensions in a NuGet package.

I expect referenced dlls to work with .Net461. 

### Issues
1. Why does netstandard2.0 not work? Why does it thinks the types are different? I have no idea what the cause of this is or how long it would take to fix.
2. Why does the NuGet package install choose netstandard2.0 over portable-net45+win8+wp8+wpa81 when dlls in the netstandard2.0 don't work and dlls int the portable-net45+win8+wp8+wpa81 do work? My guess is precedence or preference. NuGet is coded to prefer netstandard2.0 dlls over dlls in portable-net45+win8+wp8+wpa81. Is that precedence correct? Probably.
3. Since System.Threading.Tasks.Extensions dlls in the netstandard2.0 folder of the NuGet package do not work, why does the NuGet package not have a net45, net461, or such folder that will take precedence of netstandard2.0? Adding a folder for Net461 with dlls that work takes a few minutes. 

So until Issue #1 is fixed, can we get an updated NuGet package that properly points net461 projects to working dll?

**Comment by karelz** [comment]
I don't understand what you mean by "_doesn't see the System.Net.Http as the same types_". What breaks? How? Is it compile-time error or runtime error? What kind of exception?

Can you please post clear minimal steps how to reproduce the problem from a new project and what exactly happens? (let's start with simplest repro involving the least amount of other libraries and products)

**Comment by rhyous** [comment]
I mean that typeof(HttpContent) from a .Net461 project that does NOT reference System.Threading.Tasks.Extensions is NOT equal to typeof(HttpContent) from from a .Net461 project that DOES reference System.Threading.Tasks.Extensions

Error:
```
An exception of type 'System.MissingMethodException' occurred in mscorlib.dll and wasn't handled before a managed/native boundary
Additional information: Method not found: 'System.Threading.Tasks.Task`1<System.Net.Http.HttpResponseMessage> Rhyous.WebFramework.Clients.IHttpClient.PostAsync(System.String, System.Net.Http.HttpContent)'.
```

I will try to attach a project file.

**Comment by rhyous** [comment]
I keep getting an error trying to upload a zip. Probably this site has zip upload disabled. I have uploaded a zip with a minimized solution to my blog.

http://www.rhyous.com/wp-content/uploads/2018/03/Example.Bug27907.zip

**Comment by rhyous** [comment]
I've just not simplified it even further.

ProjectA - Class LIbrary .Net461 without any NuGet packages at all. 
Add One file:

```
using System;
using System.Net.Http;

namespace Example.Bug27907
{
    public class TestType
    {
        public Type HttpContentType { get { return typeof(HttpContent); } }
    }
}
```

ProjectB - Unit Test library .Net461 .
Packages:
```
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="MSTest.TestAdapter" version="1.2.0" targetFramework="net461" />
  <package id="MSTest.TestFramework" version="1.2.0" targetFramework="net461" />
  <package id="System.Threading.Tasks.Extensions" version="4.4.0" targetFramework="net461" />
</packages>
```
One file: 

```
using System.Net.Http;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Example.Bug27907.Tests
{
    [TestClass]
    public class UnitTest1
    {
        [TestMethod]
        public void TestThatThrowsTheSystemMissingMethodExceptionError()
        {
            var testType = new TestType();
            Assert.AreEqual(typeof(HttpContent), testType.HttpContentType);
        }
    }
}
```

This Unit Test must pass.

It does not.

**Comment by karelz** [comment]
I am able to reproduce it, but only with UnitTest project. Do you have a repro / scenario without UnitTest project?

**Comment by rhyous** [comment]
That is great you duplicated it! 
I didn't know it was limited to unit tests. But that is the only place I am having the problem.

This is causing a lot of head aches. Not just for me, but for many other users out there, too. Most of the developers so far have blamed unit test frameworks, such as Moq.

I am not sure of the exact cause or what causes it to happen. But further research shows the following:

With a Unit Test project, I get two loaded assemblies:
In the Unit Test that fails,
1. Put a breakpoint on the test.
2. Go to the immediate window.
3. Run this:
AppDomain.CurrentDomain.GetAssemblies()

```
AppDomain.CurrentDomain.GetAssemblies()
{System.Reflection.RuntimeAssembly[30]}
    [0]: {mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089}
    [1]: {Microsoft.TestPlatform.PlatformAbstractions, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [2]: {System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089}
    [3]: {Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [4]: {Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [5]: {System.Runtime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [6]: {System.Collections, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [7]: {System.Reflection, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [8]: {System.Runtime.Extensions, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [9]: {Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.Interface, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [10]: {System.Globalization, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [11]: {System.IO, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [12]: {Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [13]: {Microsoft.VisualStudio.TestPlatform.TestFramework.Extensions, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [14]: {System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089}
    [15]: {Example.Bug27907.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null}
    [16]: {System.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [17]: {System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089}
    [18]: {System.Reflection.Extensions, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [19]: {System.Threading, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [20]: {System.Threading.Tasks, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [21]: {System.Configuration, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [22]: {System.Xml, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089}
    [23]: {Example.Bug27907.Normal, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null}
    [24]: {System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [25]: {System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [26]: {System.Resources.ResourceManager, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [27]: {Microsoft.VisualStudio.TestPlatform.ObjectModel, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [28]: {Microsoft.TestPlatform.CoreUtilities, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [29]: {Microsoft.VisualStudio.Debugger.Runtime, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
```
1. Look at the exact loaded assemblies.
AppDomain.CurrentDomain.GetAssemblies()[24]
```
{System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [System.Reflection.RuntimeAssembly]: {System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    CodeBase: "file:///C:/Users/jbarneck/Documents/Visual Studio 2017/Projects/Example.Bug27907/Example.Bug27907.Tests/bin/Debug/System.Net.Http.DLL"
    CustomAttributes: Count = 19
    DefinedTypes: {System.RuntimeType[109]}
    EntryPoint: null
    EscapedCodeBase: "file:///C:/Users/jbarneck/Documents/Visual%20Studio%202017/Projects/Example.Bug27907/Example.Bug27907.Tests/bin/Debug/System.Net.Http.DLL"
    Evidence: {System.Security.Policy.Evidence}
    ExportedTypes: {System.Type[44]}
    FullName: "System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
    GlobalAssemblyCache: false
    HostContext: 0
    ImageRuntimeVersion: "v4.0.30319"
    IsDynamic: false
    IsFullyTrusted: true
    Location: "C:\\Users\\jbarneck\\Documents\\Visual Studio 2017\\Projects\\Example.Bug27907\\Example.Bug27907.Tests\\bin\\Debug\\System.Net.Http.dll"
    ManifestModule: {System.Net.Http.dll}
    Modules: {System.Reflection.RuntimeModule[1]}
    PermissionSet: {<PermissionSet class="System.Security.PermissionSet"
version="1"
Unrestricted="true"/>
}
    ReflectionOnly: false
    SecurityRuleSet: Level2
```
1. AppDomain.CurrentDomain.GetAssemblies()[25]

```
{System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [System.Reflection.RuntimeAssembly]: {System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    CodeBase: "file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Net.Http/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Net.Http.dll"
    CustomAttributes: Count = 23
    DefinedTypes: {System.RuntimeType[102]}
    EntryPoint: null
    EscapedCodeBase: "file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Net.Http/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Net.Http.dll"
    Evidence: {System.Security.Policy.Evidence}
    ExportedTypes: {System.Type[44]}
    FullName: "System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
    GlobalAssemblyCache: true
    HostContext: 0
    ImageRuntimeVersion: "v4.0.30319"
    IsDynamic: false
    IsFullyTrusted: true
    Location: "C:\\WINDOWS\\Microsoft.Net\\assembly\\GAC_MSIL\\System.Net.Http\\v4.0_4.0.0.0__b03f5f7f11d50a3a\\System.Net.Http.dll"
    ManifestModule: {System.Net.Http.dll}
    Modules: {System.Reflection.RuntimeModule[1]}
    PermissionSet: {<PermissionSet class="System.Security.PermissionSet"
version="1"
Unrestricted="true"/>
}
    ReflectionOnly: false
    SecurityRuleSet: Level2
```

So the unit test project is somehow loading a 4.0.0.0 assembly from the GAC.

In a non-unit test project. The assembly of version 4.0.0.0 from the GAC is never loaded. It is always the 4.2.0.0 version.

```
AppDomain.CurrentDomain.GetAssemblies()
{System.Reflection.RuntimeAssembly[5]}
    [0]: {mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089}
    [1]: {Example.Bug27907.Different, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null}
    [2]: {Example.Bug27907.Normal, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null}
    [3]: {System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
    [4]: {Microsoft.VisualStudio.Debugger.Runtime, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
```

In a Unit Test without System.Threading.Tasks.Extensions installed, it always loads the 4.0.0.0 version from the GAC.

Now the Properties (highlight the dll in references and press F4) are different too.

Path to Dll in Unit Test Project with NuGet package: 
```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Microsoft\Microsoft.NET.Build.Extensions\net461\lib\System.Net.Http.dll
```
Path to dll in Unit Test Project with no NuGet Packages:
```
C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.1\System.Net.Http.dll
```

So basically, the GAC is 4.0.0.0 and so is the file in Reference Assemblies, but the other files are 4.2.0.0.

The Unit Test with the NuGet package, loads 4.2.0.0 first, then when it calls the related assembly, the related assembly loads up 4.0.0.0 and the two types are from different versions of assemblies.

The Unit Test without the NuGet package, loads 4.0.0.0 first, then when it calls the related assembly, the related assembly loads up 4.0.0.0 and the two types are from different versions of assemblies.

Strangely, the projects that aren't unit test projects seem fine. They both load 4.2.0.0 always.

**Comment by karelz** [comment]
> Not just for me, but for many other users out there, too

Can you point me to some evidence showing the impact? That would help us prioritize a fix.

The unit test project is missing bindingRedirects, that's why there are 2 assemblies loaded leading to type inequality. @joperezr can you please help find out what's wrong in this scenario? See simple repro steps above - https://github.com/dotnet/corefx/issues/27907#issuecomment-371942153 (4.6.1 library + 4.6.1 UnitTest project, nothing more).

**Comment by karelz** [comment]
@joperezr you mentioned this is turned on by default in newer VS, can you please chime in more details?

**Comment by karelz** [comment]
The overall end-to-end scenario is worrisome - cc @AlexGhiondea @weshaggard

**Comment by joperezr** [comment]
>  newer VS, can you please chime in more details?

I'm not 100% sure but I believe that SDK projects have this turned on by default. I'll double check.

**Comment by karelz** [comment]
From a brief chat with @AlexGhiondea it also seems that running on 4.7.2 (which is in Early access preview) will not hit the problem. Given that it is developer-only scenario which is broken, it should be acceptable workaround (and it will "fix itself" going forward).

@rhyous we are still interested to hear more details about the impact of this specific problem. To our knowledge it is not that widespread (we've got only a couple of reports on it). If we're missing some channels / feedback, we would like to know that ASAP. Thanks!

**Comment by rhyous** [comment]
Ahh...Priority discussion.

So for me, it cost 1 week of dev time and it is still costing me time. We couldn't deploy because unit tests are not passing. We couldn't make our unit tests pass until I figured this out. I beat my head against it for a while.  We found a workaround, to modify the .csproj directly. That cost is in the past, so it is not a future cost.

We use HttpClient heavily. So many of our tests use Moq an will fail without the workaround. However, the workaround to modify the csproj for dozen of test projects mostly has us working.

But then this week, another developer updated the NuGet package locally and it over-wrote the custom csproj edit. He lost a day trying to test and didn't know why his unit tests were failing. So he couldn't check in his  projects, etc. So my team lost another day.

Yet another developer started a new test project this week, and so my csproj workaround was not already in place. He wasted a day before I found out and helped him. So we lost another day.

So yes, this bug is still costing us time.

Also, it appears to be affecting more than just me. It is affecting the subset of the Moq community that is testing anything from System.Net.Http. How big that subset is, I don't know. It seems that Moq is the most used C# mocking framework for unit tests. Because it is limited to unit tests that cover the System.Net.Http library, I would make a guess that it affects less than 10% of Moq users. But of course, I have no data do back that up other than this bug which has other users complaining: https://github.com/moq/moq4/issues/566

So for me, the priority is waning from fix it right now, to whenever you deploy next is fine. 

However, you might ask the Moq team how this is affecting them.

**Comment by petedavis** [comment]
Having the same issue now and lost a day before ending up here. Using System.Reactive 4.1 takes a dependency on System.Threading.Tasks.Extensions 4.5.1.  Using Net462 I am getting all files from `C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Microsoft\Microsoft.NET.Build.Extensions\net461\lib` copied into my build output folder.

This is because when installed the hint path is:
`<HintPath>..\packages\System.Threading.Tasks.Extensions.4.5.1\lib\netstandard2.0\System.Threading.Tasks.Extensions.dll</HintPath>`

Given that the package states that for net 4.5+ the only dependency should be `System.Runtime.CompilerServices.Unsafe`

Simple to reproduce. Create a new Net Framework project targeting framework 4.6.2. Install System.Threading.Tasks.Extensions version 4.3.0 and there will only be two dll's in the build folder.

Update to 4.4.0+ and there will be over 100 files for a package that states it has a single dependency.

**Comment by DMEvans** [comment]
I too have lost about a day on this particular issue until I finally found my way to this thread.  Is there any permanent solution on the horizon for this?

**Comment by kevbry** [comment]
Same here. A day lost.

**Comment by joperezr** [comment]
If you are using an updated version of VS, I would expect that the extra files that get copied to your output folder along with some extra binding redirects that will be added to your app.config file to fix all the MissingMethodExceptions. Could someone hitting this attach the simple repro to the issue so we can validate that an updated version of VS fixes this?

**Comment by rhyous** [comment]
@joperezr I just saw this. I have a link to a sample project reproducing this issue: https://github.com/dotnet/corefx/issues/27907#issuecomment-371940323

**Comment by shashank198616** [comment]
I was facing the same problem
Error: System.MissingMethodException : Method not found (on System.Net.Http.HttpContent)
My tests were running locally but were failing during Build. Adding below to my config file worked and the build passed.
<dependentAssembly>
<assemblyIdentity name="System.Net.Http" publicKeyToken="your token number" culture="neutral" />
<bindingRedirect oldVersion="0.0.0.0-4.2.0.0" newVersion="4.0.0.0" />
</dependentAssembly>

**Comment by joperezr** [comment]
@shashank198616 even though this seems to have fixed the problem that is not the right fix, since you are essentially downgrading the version of System.Net.Http that your app gets at runtime, so you may get a version that is missing Api required by your application. The right fix will depend on your project, but generally the advice is to 1) Enable binding-redirect generation, 2) Update VS to get latest changes in MSBuild, 3) Migrate projects out of packages.config and into PackageReference. The combination of those 3 usually fix these type of issues. Another thing that will make things work as well is to retarget to 4.7.2

**Comment by shashank198616** [comment]
That makes sense, as I knew this would not be a permanent solution. My
tests were passing locally while they failed during build. Can you please
elaborate on below so I could understand better, as I could not re-produce
this locally, as they were failing during build.

 1) Enable binding-redirect generation,
 2) Update VS to get latest changes in MSBuild, (are you recommending to
update to 2019, as mine is 2017 already)?
 3) Migrate projects out of packages.config and into PackageReference-
reference to latest version?

**Comment by joperezr** [comment]
@shashank198616 sure.

1) You can find more information on how to enable automatic binding redirects here: https://docs.microsoft.com/en-us/dotnet/framework/configure-apps/how-to-enable-and-disable-automatic-binding-redirection 
That link should give you info on how to do it from VS, and also if you want to manually do it in your project file.

2) Latest 2017 with all updates should be good enough. All of the changes that we have made into the MSBuild engine for this issue are present on the latest update of 2017.

3) Not all projects support it yet, but try to right-click your project References (or packages.config) in the solution explorer, and see if there is an option to "Migrate packages.config to PackageReference..." that should look like this:
![image](https://user-images.githubusercontent.com/13854455/56235911-3d589d80-603d-11e9-96f9-ce6e99a0ccc5.png)
After clicking that, you should now no longer have packages.config and instead have the references built in to your csproj as PackageReference. This is the new NuGet model for references, and fixes a wide number of issues very similar to the ones this thread is talking about.
